### DISCUSSION TITLE
package.json engines ignores by pnpm

### DISCUSSION THREAD

**Discussion by tictac22** [pnpm/pnpm#4811] [Q&A]
I want that in my project everyone uses pnpm.
I set in `package.json engines` such fields
```
"engines": {
		"npm": "please-use-pnpm",
		"yarn": "please-use-pnpm",
		"node": ">=18.0.0",
		"pnpm": ">=7.0.0"
}
```
When I try to `yarn install` terminal throws an error. So everything work perfectly.
But with `npm install` it ignores `engines npm` and tries to install  `node_modules`.
How can it be fixed?
Node version: `18.2.0`
pnpm version: `7.1.6`

**Comment by zkochan** [comment]
We cannot change how npm works. If you enable corepack and add the packageManager field to your package.json, then it will prevent others from running npm. But only when corepack is enable on their computer.

**Reply by tictac22** [reply]
Am I doing it right?
`corepack enable`
`corepack prepare pnpm@7.1.6 --activate`
and in package.json `"packageManager": "pnpm@7.1.6",`.
And still when I type `yarn install` terminal thorws an error `This project is configured to use pnpm`.
But with `npm install` it doesn't throw the error and tries to install node_modules.

**Reply by zkochan** [reply]
In that case, there is no solution.

**Reply by Aghassi** [reply]
Corepack only supports `yarn` and `pnpm`. In addition, I'm pretty sure it only supports newer versions of `yarn`, not `yarn classic`

**Comment by r-moore** [comment]
I use this little preinstall script:
```
"scripts":{
  "preinstall": "node -e \"if(process.env.npm_execpath.indexOf('pnpm') === -1) throw new Error('You must use PNPM')\""
}
```

**Reply by Aghassi** [reply]
This works beautifully and doesn't depend on anything. Easily customizable too. Thank you!

**Reply by zkochan** [reply]
This will only stop you from running `npm install`. `npm install something` will still work.

**Reply by Aghassi** [reply]
This is a good call out. In our case, we are trying to just teach people that a workspace we have requires `pnpm i` instead of `yarn install` as there are other areas still using yarn :).
